index.rst

```diff
@@ -1,25 +1,429 @@
+Protobuf also was designed with schema evolution in mind.
+It uses manually numbered fields, and as long as you don’t re-use a number you can add new fields and delete old ones.
+
+json schema was not designed for schema evolution, but Kafka does support it, as `described here <https://yokota.blog/2021/03/29/understanding-json-schema-compatibility/>`_.
+I tested schema evolution and found, as advertised in that link, setting additionalProperties#false allows one to add fields, but not delete them.
+Setting additionalProperties#false our preferred way to define schemas,  because it detects typos in field names.
+However, the inability to delete fields is a drawback.
+
+Arrays
+------
+
+SAL arrays are all fixed-length.
+This turns out to be a significant issue because neither Avro nor protobuf supports any kind of constraint on array length.
+
+json schema is the only choice that supports array length constraints, and we would have to test if Kafka enforces them.
+In order to constrain array lengths using Avro or Protobuf we will have to manually check the data ourselves.
+This is practical as long as we retain additional information (beyond that used to generate the schema used by Kafka) when we parse the XML files that define our SAL interfaces.
+
+Scalar Data Types
+-----------------
+
+Avro supports all data types we use except unsigned integers.
+Specifically Avro supports null, boolean, int (32 bit), long (64 bit), float, double, bytes (sequence of 8-bit unsigned values), and string (unicode). Note that SAL does not support null or bytes.
+We can certainly live without unsigned integers, since we have been doing so for feeding the EFD.
+
+Protobuf supports all data types we use.
+Specifically it supports bool, float, double, int32, int64, uint32, uint64 (and several variants of these with different encoding), float, double, bytes, string.
+
+json schema has very weak support for numeric data types, as it only supports integers and float.
+Specifically it supports null, boolean, integer, number (float or int), string.
+This is a major limitation.
+Between this and the lack of support for evolving schemas by removing fields, I believe json schema is not a suitable choice for us.
+
+Default Float Value
+-------------------
+
+The default value for float is a tricky question.
+NaN is most principled, but it has issues:
+
+* For Avro: NaN is not yet supported as a default in C++, though there is a `patch <https://issues.apache.org/jira/browse/AVRO-2880>`_ that may be accepted.
+  If we want to use NaN as the default float value then we must make sure C++, Java, and Python all support it.
+
+* Protobuf has a fixed default value for each type; float the default is 0.
+
+* Json schema is like Avro in supporting per-field default values.
+  However, I don’t know if Kafka uses the specified default values (the standard json schema validators do not), and if Kafka allows NaN (json does not, but extensions often do and allow it to be spelled as "nan").
```

Comment:
> However, it is not clear if Kafka uses the specified default values (the standard json schema validators do not), and if Kafka allows NaN (json does not, but extensions often do and allow it to be spelled as "nan").

Reply:
I went with:

Support for Google protobuf, and json schema is recent.
Information is very sparse about how Kafka uses json schema.
For example: does Kafka fill in default values? The standard json schema validators do not.